README

README for Games-Bowling-Scorecard 0.103

This modules lets you turn this:

  qw(6 1 7 2 10 9 1 10 10 10 0 0 8 2 10 9 1);

Into this:

    +-----+-----+-----+-----+-----+-----+-----+-----+-----+-------+
    | 6 1 | 7 2 | X   | 9 / | X   | X   | X   | - - | 8 / | X 9 / |
    |   7 |  16 |  36 |  56 |  86 | 106 | 116 | 116 | 136 |   156 |


Changes
Changelog for Games-Bowling-Scorecard

{{$NEXT}}

0.106     2022-11-25 18:58:53-05:00 America/New_York
        - update dist metadata
        - AsText is now nicely box-drawn; AsASCII is just ASCII
        - you can record a split, which looks nice in non-ASCII form

0.105     2013-09-24 20:56:26 America/New_York
          replace a busted-on-Win32 test file

0.104     2013-08-24 08:19:51 America/New_York
          repackage!  now using recent Dist::Zilla and github

0.103     2008-06-02
          experiment in dist packaging: this dist not built by EUMM, MI, or MB

0.102     2008-05-27
          experiment in dist packaging: no MANIFEST included

0.101     2007-02-22
          add README (for kwalitee)

0.100     2007-01-22
          add roll_ok checks
          improve test coverage
          fix a number of AsText bugs for tenth frame
          improve tests for AsText

0.011     2007-01-21
          fix version code/doc mismatches
          add score_through method
          add AsText to produce ASCII scorecards

0.001     2007-01-21
          first release!


lib/Games/Bowling/Scorecard.pm
use v5.24.0;
use warnings;
package Games::Bowling::Scorecard;
# ABSTRACT: score your bowling game easily

=head1 SYNOPSIS

  use Games::Bowling::Scorecard;

  my $card = Games::Bowling::Scorecard->new;

  $card->record(6,1);  # slow start
  $card->record(7,2);  # getting better
  $card->record(10);   # strike!
  $card->record(9,1);  # picked up a spare
  $card->record(10) for 1 .. 3; # turkey!
  $card->record(0,0);  # clearly distracted by something
  $card->record(8,2);  # amazingly picked up 7-10 split
  $card->record(10, 9, 1); # pick up a bonus spare

  printf "total score: %u\n", $card->score; # total score: 156, lousy!

=head1 DESCRIPTION

Scoring ten-pin bowling can be confusing for new players.  Frames can't always
be scored until several frames later, and then there's that weird tenth frame.
Modern bowling alleys incorporate computer scoring into the pin cleanup
mechanism, so it's easy to just concentrate on throwing a perfect game and not
on grease-pencilling the sheet for the overhead.

What's one to do, though, when bowling cantaloupes at beer bottles in one's
back yard?  Now, with Games::Bowling::Scorecard, it's easy to improvise a
scoring device -- maybe on a mobile phone running Symbian Perl.

=cut

use Games::Bowling::Scorecard::Frame;

=method new

This returns a new scorecard object.  It does not take any arguments.

=cut

sub new {
  my ($class) = @_;

  my $self = bless { frames => [ ] } => $class;

  return $self;
}

=method frames

  my @frames = $card->frames;

This method returns all of the frames for the game.  This will return all
frames in which scores have been recorded, and possibly one final frame with no
recorded balls.  It will never return any frames after that.

Frames are returned as Games::Bowling::Scorecard::Frame objects.

=cut

sub frames {
  my ($self) = @_;

  return @{ $self->{frames} };
}

=method current_frame

The current frame is the frame into which the next ball will be recorded.  If
the card is done, this method returns false.

=cut

sub current_frame {
  my ($self) = @_;

  return if $self->is_done;

  my @frames = $self->frames;

  my $frame = pop @frames;

  return $self->_next_frame if !$frame || $frame->is_done;

  return $frame;
}

sub _next_frame {
  my ($self) = @_;

  my $frame = $self->frames == 9
            ? do {
                require Games::Bowling::Scorecard::Frame::TenPinTenth;
                Games::Bowling::Scorecard::Frame::TenPinTenth->new;
              }
            : Games::Bowling::Scorecard::Frame->new;

  push @{ $self->{frames} }, $frame;

  return $frame;
}

=method pending_frames

This method returns any completed frames the score of which has not yet been
finalized.  This includes spares and strikes, before the next ball or balls
have been recorded.

=cut

sub pending_frames {
  my ($self) = @_;

  my @pending_frames = grep { $_->is_pending } $self->frames;
}

=method record

  $card->record(@balls);

This method makes a record of a ball or balls.  It is passed a list of bowling
results, each being a number of pins knocked down by the ball.

For example:

  $card->record(0, 0);  # two gutter balls

  $card->record(6, 4);  # a spare

  $card->record( (0, 0) x 10); # the worst game you could play

  $card->record( (10) x 12 ); # a perfect game

An exception will be raised if this method is called on a scorecard that's
done.

If you need to record a ball with more arguments, you can pass them together in
an array reference.  For example, to pick up an incredible 7-10 split, you
might call:

  $card->record([ 8, { split => 1 } ], 2);

The first ball records that it's a split, and the second ball just gets two
pins.

=cut

sub record { ## no critic Ambiguous
  my $self  = shift;
  my @balls = @_;

  for my $i (0 .. $#balls) {
    Carp::croak "can't record more balls on a completed scorecard"
      if $self->is_done;

    my ($ball, $arg) = ref $balls[$i]
                     ? ($balls[$i][0], $balls[$i][1])
                     : ($balls[$i]);

    for my $pending ($self->pending_frames) {
      $pending->record($ball);
    }

    $self->current_frame->record($ball, $arg);
  }
}

=method score

This method returns the current score.  It will include the tentative score for
all pending frames.

=cut

sub score {
  my ($self) = @_;

  my $score = 0;
  $score += $_->score for $self->frames;

  return $score;
}

=method score_through

  my $score = $card->score_through($n)

This method returns the score as of the end of the I<n>th frame.  If that
frame's cannot be definitively stated, because it is pending or not done, undef
is returned.

=cut

sub score_through {
  my ($card, $n) = @_;

  Carp::croak "frame out of range" unless $n >= 1 and $n <= 10;

  my @frames = $card->frames;
  my $score = 0;

  INDEX: for my $idx (0 .. $n - 1) {
    my $frame = $frames[ $idx ];
    return undef if $frame->is_pending or not $frame->is_done;

    $score += $frame->score;
  }

  return $score;
}

=method is_done

This returns true if the scorecard is done.  The scorecard is done if its
contents indicate that the player's game is over.

=cut

sub is_done {
  my ($self) = @_;

  my @frames = $self->frames;

  return (@frames == 10 and $frames[9]->is_done);
}

=head1 TODO

=for :list
* maybe a way to indicate a split

=head1 SECRET ORIGINS

In late 2006, I hadn't bowled in something like ten years.  I got a Wii, and
while I recognized the little triangle and X marks on the Wii Sports Bowling
scorecard, I couldn't remember how on earth scoring worked.  Once I thought I
had a handle on it, I thought writing this would be a good way to cement it in
my mind.

=cut

300;


lib/Games/Bowling/Scorecard/Frame.pm
use v5.24.0;
use warnings;

package Games::Bowling::Scorecard::Frame;
# ABSTRACT: one frame on a scorecard

=head1 DESCRIPTION

A frame is one attempt to knock down all ten pins -- unless it's the tenth
frame, in which case it's so goofy that you need to use a different class,
L<Games::Bowling::Scorecard::Frame::TenPinTenth>.  A frame is done when you've
bowled twice or knocked down all the pins, and it's pending until its score can
be definitively be stated.

=cut

use Carp ();

=method new

This method returns a new frame object.

=cut

sub new {
  my ($class) = @_;

  bless {
    balls => [],
    score => 0,

    done    => 0,
    pending => 0,
  } => $class;
}

=method record

  $frame->record($ball, \%arg);

This method records a single ball against the frame.  This method is used for
both the current frame and for pending frames.  It updates the frame's score
and whether the frame is done or pending.

The only valid argument in C<%arg> is C<split>. If true, it indicates the pins
are split.  This can only be passed on the first ball of a frame.

=cut

sub _assert_split_ok {
  my ($self, $ball) = @_;

  if ($self->{balls}->@*) {
    Carp::croak "can't record a split on second ball in a frame";
  }

  if ($ball >= 9) {
    Carp::croak "you can't split if you knocked down $ball pins!";
  }

  return;
}

sub record { ## no critic Ambiguous
  my ($self, $ball, $arg) = @_;

  if ($arg->{split}) {
    $self->_assert_split_ok($ball);

    $self->{split} = 1;
  }

  if ($self->is_done) {
    if ($self->is_pending) {
      $self->{pending}--;
      $self->{score} += $ball;
      return;
    } else {
      Carp::croak "two balls already recorded for frame";
    }
  }

  $self->roll_ok($ball);

  push @{ $self->{balls} }, $ball;
  $self->{score} += $ball;

  $self->_check_done;
  $self->_check_pending;
}

sub was_split {
  return $_[0]->{split} ? 1 : 0;
}

sub _check_pending {
  my ($self) = @_;
  return unless $self->is_done;

  my @balls = $self->balls;

  return $self->{pending} = 2 if @balls == 1 and $balls[0] == 10;
  return $self->{pending} = 1 if @balls == 2 and $balls[0] + $balls[1] == 10;
}

sub _check_done {
  my ($self) = @_;

  my @balls = $self->balls;

  $self->{done} = 1 if (@balls == 1 and $balls[0] == 10) or @balls == 2;
}

=method roll_ok

  $frame->roll_ok($ball);

This method asserts that given value is an acceptable number to score next in
this frame.  It checks that:

  * the frame is not already done
  * $ball is defined, an integer, and between 0 and 10
  * $ball would not bring the total number of pins downed above 10

=cut

sub roll_ok {
  my ($self, $ball) = @_;

  Carp::croak "the frame is done" if $self->is_done;
  Carp::croak "you can't bowl an undefined number of pins!" if !defined $ball;
  Carp::croak "you can't bowl more than 10 on a single ball" if $ball > 10;
  Carp::croak "you can't bowl less than 0 on a single ball" if $ball < 0;
  Carp::croak "you can't knock down a partial pin" if $ball != int($ball);

  my $i = 0;
  $i += $_ for $self->balls, $ball;

  Carp::croak "bowling a $ball would bring the frame above 10" if $i > 10;
}

=method score

This method returns the current score for the frame, even if the frame is not
done or is pending further balls.

=cut

sub score {
  my ($self) = @_;
  return $self->{score};
}

=method is_pending

This method returns true if the frame is pending more balls -- that is, it
returns true for strikes or spares which have not yet recorded the results of
subsequent balls.

=cut

sub is_pending {
  my ($self) = @_;
  return $self->{pending};
}

=method is_done

This method returns true if the frame is done.

=cut

sub is_done {
  my ($self) = @_;
  return $self->{done};
}

=method balls

This method returns the balls recorded against the frame, each ball returned as
the number of pins it knocked down.  In scalar context, it returns the number
of balls recoded against the frame.

=cut

sub balls {
  my ($self) = @_;
  return @{ $self->{balls} };
}

300;


lib/Games/Bowling/Scorecard/AsASCII.pm
use v5.24.0;
use warnings;
package Games::Bowling::Scorecard::AsASCII;
# ABSTRACT: format a bowling scorecard as ASCII text

=head1 SYNOPSIS

  use Games::Bowling::Scorecard;

  my $card = Games::Bowling::Scorecard->new;

  $card->record(6,1);  # slow start
  $card->record(7,2);  # getting better
  $card->record(10);   # strike!
  $card->record(9,1);  # picked up a spare
  $card->record(10) for 1 .. 3; # turkey!
  $card->record(0,0);  # clearly distracted by something
  $card->record(8,2);  # amazingly picked up 7-10 split
  $card->record(10, 9, 1); # pick up a bonus spare

  print Games::Bowling::Scorecard::AsText->card_as_text($card);

The above outputs:

  +-----+-----+-----+-----+-----+-----+-----+-----+-----+-------+
  | 6 1 | 7 2 | X   | 9 / | X   | X   | X   | - - | 8 / | X 9 / |
  |   7 |  16 |  36 |  56 |  86 | 106 | 116 | 116 | 136 |   156 |

=head1 WARNING

This module's interface is almost certain to change, whenever the author gets
around to it.

=head1 DESCRIPTION

So, you've written a bowling record-keeper and now you want to print out
scorecards to your dynamic Gopher site.  Games::Bowling::Scorecard has taken
care of the scoring, but now you need to worry about all those slashes and
dashes and X's

=method card_as_text

  my $text = Games::Bowling::Scorecard::AsText->card_as_text($card);

Given a scorecard, this method returns a three-line text version of the card,
using standard notation.  A total is kept only through the last non-pending
frame.

=cut

use Carp ();

sub card_as_text {
  my ($self, $card) = @_;

  my $hdr = '+-----+-----+-----+-----+-----+-----+-----+-----+-----+-------+';
  my $balls  = '';
  my $scores = '';

  my @frames = $card->frames;
  INDEX: for my $i (0 .. 8) {
    my $frame = $frames[ $i ];
    unless ($frame) {
      $_  .= '|     ' for $balls, $scores;
      next INDEX;
    }

    $balls .= sprintf '| %s ', $self->_two_balls($frame->balls);

    my $score = $card->score_through($i + 1);
    $scores .= defined $score
             ? sprintf '| %3u ', $score
             : '|     ';
  }

  TENTH: for (1) {
    my $frame = $frames[ 9 ];

    unless ($frame) {
      $_ .= '|       |' for $balls, $scores;
      last TENTH;
    }

    $balls .= sprintf '| %s |', $self->_three_balls($frame->balls);

    my $score = $card->score_through(10);

    $scores .= defined $score
             ? sprintf '|   %3u |', $score
             : '|       |';
  }

  return "$hdr\n"
       . "$balls\n"
       . "$scores\n";
}

sub _two_balls {
  my ($self, $b1, $b2) = @_;

  return '   ' unless defined $b1;

  sprintf '%s %s',
    $b1 == 10 ? 'X' : $b1 || '-',
    $b1 == 10 ? ' ' : defined $b2 ? $b1 + $b2 == 10 ? '/' : $b2 || '-' : ' ';
}

sub _three_balls {
  my ($self, $b1, $b2, $b3) = @_;

  return '     ' unless defined $b1;

  if ($b1 == 10) {
    return 'X    ' unless defined $b2;

    return sprintf 'X X %s', defined $b3 ? $b3 == 10 ? 'X' : $b3 || '-' : ' '
      if $b2 == 10;

    return sprintf 'X %s', $self->_two_balls($b2, $b3);
  } elsif (not defined $b2) {
    return sprintf '%s    ', $b1 || '-';
  } elsif ($b1 + $b2 == 10) {
    return sprintf '%s %s',
      $self->_two_balls($b1, $b2),
      defined $b3 ? $b3 == 10 ? 'X' : $b3 || '-' : ' ';
  } else {
    return sprintf '%s  ', $self->_two_balls($b1, $b2);
  }
}

300;


lib/Games/Bowling/Scorecard/Frame/TenPinTenth.pm
use v5.24.0;
use warnings;

package Games::Bowling::Scorecard::Frame::TenPinTenth;
use parent qw(Games::Bowling::Scorecard::Frame);
# ABSTRACT: ten pin's weird 10th frame

=head1 DESCRIPTION

The tenth frame of ten pin bowling is weird.  If you bowl a strike or spare,
you're allowed to throw balls to complete the frame's scoring.  The extra balls
are only counted for bonus points.  In other words, if the first two balls in
the tenth frame are strikes, the second ball is not counted as a "pending"
strike.  If this is confusing, don't worry!  That's why you're using this
module.

=method is_done

The tenth frame is done if: (a) three balls have been bowled or (b) two balls
have been bowled, totalling less than ten.

=cut

sub is_done {
  my ($self) = @_;

  my @balls = $self->balls;

  return 1 if @balls == 3 or @balls == 2 and $balls[0] + $balls[1] < 10;
  return;
}

=method is_pending

The tenth frame is never pending.  Once it's done, its score is final.

=cut

sub is_pending {
  return 0;
}

=method roll_ok

The tenth frame's C<roll_ok> is identical to the standard C<roll_ok>, but
replaces the "can't total more than 10" rule with a more complex rule.

=cut

sub roll_ok {
  my ($self, $ball) = @_;

  eval { $self->SUPER::roll_ok($ball) };

  if (my $error = $@) {
    return if $error =~ /would bring the frame above 10/;
    die $error;
  }
}

sub _assert_split_ok {
  my ($self, $ball) = @_;

  if ($self->{balls}->@* > 2) {
    Carp::croak "can't record a split on third ball in tenth frame";
  }

  if ($ball >= 9) {
    Carp::croak "you can't split if you knocked down $ball pins!";
  }

  return;
}


300;


lib/Games/Bowling/Scorecard/AsText.pm
use v5.24.0;
use warnings;
package Games::Bowling::Scorecard::AsText;
# ABSTRACT: format a bowling scorecard as text

use utf8;

=encoding utf8

=head1 SYNOPSIS

  use Games::Bowling::Scorecard;

  my $card = Games::Bowling::Scorecard->new;

  $card->record(6,1);  # slow start
  $card->record(7,2);  # getting better
  $card->record(10);   # strike!
  $card->record(9,1);  # picked up a spare
  $card->record(10) for 1 .. 3; # turkey!
  $card->record(0,0);  # clearly distracted by something
  $card->record(8,2);  # amazingly picked up 7-10 split
  $card->record(10, 9, 1); # pick up a bonus spare

  print Games::Bowling::Scorecard::AsText->card_as_text($card);

The above outputs:

  ┏━━━━━┳━━━━━┳━━━━━┳━━━━━┳━━━━━┳━━━━━┳━━━━━┳━━━━━┳━━━━━┳━━━━━━━┓
  ┃ 6 1 ┃ 7 2 ┃ ╳   ┃ 9 ◢ ┃ ╳   ┃ ╳   ┃ ╳   ┃ - - ┃ 8 ◢ ┃ ╳ 9 ◢ ┃
  ┃   7 ┃  16 ┃  36 ┃  56 ┃  86 ┃ 106 ┃ 116 ┃ 116 ┃ 136 ┃   156 ┃
  ┗━━━━━┻━━━━━┻━━━━━┻━━━━━┻━━━━━┻━━━━━┻━━━━━┻━━━━━┻━━━━━┻━━━━━━━┛

=head1 WARNING

This module's interface is almost certain to change, whenever the author gets
around to it.

=head1 DESCRIPTION

So, you've written a bowling record-keeper and now you want to print out
scorecards to your dynamic Gopher site.  Games::Bowling::Scorecard has taken
care of the scoring, but now you need to worry about all those slashes and
dashes and X's

=method card_as_text

  my $text = Games::Bowling::Scorecard::AsText->card_as_text($card);

Given a scorecard, this method returns a three-line text version of the card,
using standard notation.  A total is kept only through the last non-pending
frame.

=cut

use Carp ();

sub card_as_text {
  my ($self, $card) = @_;

  my $hdr = '┏━━━━━┳━━━━━┳━━━━━┳━━━━━┳━━━━━┳━━━━━┳━━━━━┳━━━━━┳━━━━━┳━━━━━━━┓';
  my $ftr = '┗━━━━━┻━━━━━┻━━━━━┻━━━━━┻━━━━━┻━━━━━┻━━━━━┻━━━━━┻━━━━━┻━━━━━━━┛';
  my $balls  = '';
  my $scores = '';

  my @frames = $card->frames;
  INDEX: for my $i (0 .. 8) {
    my $frame = $frames[ $i ];
    unless ($frame) {
      $_  .= '┃     ' for $balls, $scores;
      next INDEX;
    }

    $balls .= sprintf '┃ %s ',
      $self->_two_balls($frame->was_split, $frame->balls);

    my $score = $card->score_through($i + 1);
    $scores .= defined $score
             ? sprintf '┃ %3u ', $score
             : '┃     ';
  }

  TENTH: for (1) {
    my $frame = $frames[ 9 ];

    unless ($frame) {
      $_ .= '┃       ┃' for $balls, $scores;
      last TENTH;
    }

    $balls .= sprintf '┃ %s ┃',
      $self->_three_balls($frame->was_split, $frame->balls);

    my $score = $card->score_through(10);

    $scores .= defined $score
             ? sprintf '┃   %3u ┃', $score
             : '┃       ┃';
  }

  return "$hdr\n"
       . "$balls\n"
       . "$scores\n"
       . "$ftr\n";
}

my @NUM = qw(
  - 1 2 3 4 5 6 7 8 9
    ① ② ③ ④ ⑤ ⑥ ⑦ ⑧ ⑨ 
);

sub _two_balls {
  my ($self, $split, $b1, $b2) = @_;

  return '   ' unless defined $b1;

  my $c = $NUM[ (9 * $split) + $b1 ];

  sprintf '%s %s',
    $b1 == 10 ? '╳' : $c,
    $b1 == 10 ? ' ' : defined $b2 ? $b1 + $b2 == 10 ? '◢' : $b2 || '-' : ' ';
}

sub _three_balls {
  my ($self, $split, $b1, $b2, $b3) = @_;

  return '     ' unless defined $b1;

  if ($b1 == 10) {
    return '╳    ' unless defined $b2;

    return sprintf '╳ ╳ %s', defined $b3 ? $b3 == 10 ? '╳' : $b3 || '-' : ' '
      if $b2 == 10;

    return sprintf '╳ %s', $self->_two_balls($split, $b2, $b3);
  } elsif (not defined $b2) {
    return sprintf '%s    ', $b1 || '-';
  } elsif ($b1 + $b2 == 10) {
    return sprintf '%s %s',
      $self->_two_balls($split, $b1, $b2),
      defined $b3 ? $b3 == 10 ? '╳' : $b3 || '-' : ' ';
  } else {
    return sprintf '%s  ', $self->_two_balls($split, $b1, $b2);
  }
}

300;
